sources/engine/Xenko.Physics/Engine/PhysicsComponent.cs

```diff
@@ -449,70 +460,47 @@ public void RemoveDebugEntity(Scene scene)
         /// Computes the physics transformation from the TransformComponent values
         /// </summary>
         /// <returns></returns>
-        internal void DerivePhysicsTransformation(out Matrix derivedTransformation)
+        internal void DerivePhysicsTransform(ref Matrix fromMatrix, out Matrix outMatrix)
         {
-            Matrix rotation;
-            Vector3 translation;
-            Vector3 scale;
-            Entity.Transform.WorldMatrix.Decompose(out scale, out rotation, out translation);
+            fromMatrix.Decompose(out Vector3 scale, out Matrix rotation, out Vector3 translation);
+            DerivePhysicsTransform(translation, rotation, scale, out outMatrix);
+        }
 
-            var translationMatrix = Matrix.Translation(translation);
-            Matrix.Multiply(ref rotation, ref translationMatrix, out derivedTransformation);
+        internal void DerivePhysicsTransform(Vector3? worldPosition, Matrix? worldRotation, Vector3? worldScale, out Matrix outMatrix)
```

Comment:
Curious, wouldn't it be nicer to work with `Quaternion`  here?

Reply:
Yes, but Decompose outputs a rotation matrix, not a quaternion. To avoid the conversion, I just work with the Matrix.

Reply:
There's an overload of `Matrix.Decompose` which works with `Quaternion`:
https://github.com/xenko3d/xenko/blob/releases/3.0.0.6/sources/core/Xenko.Core.Mathematics/Matrix.cs#L702

Reply:
Yes, but that does some extra math to do the conversion, which can be avoided internally if I just work with a Matrix (if I'm understanding this correctly).